R/dob_from_chi.R

```diff
@@ -0,0 +1,200 @@
+  ))
+
+  new_na_count <- sum(is.na(guess_dob)) - na_count
+
+  if (new_na_count > 0) {
+    message(glue::glue("{format(new_na_count, big.mark = ',')} CHI number{ifelse(new_na_count > 1, 's produced ambiguous dates', ' produced an ambiguous date')} and will be given NA for DoB, if possible try different values for min_date and/or max_date"))
+  }
+
+
+  return(guess_dob)
+}
+
+#' @title Extract age from the CHI number
+#'
+#' @description \code{age_from_chi} takes a CHI number or a vector of CHI numbers
+#' and returns the age as implied by the CHI number(s). If the DoB is ambiguous
+#' it will return NA. It uses \code{dob_from_chi}.
+#'
+#' @param chi_number a CHI number or a vector of CHI numbers with \code{character} class.
+#' @param ref_date calculate the age at this date, default is to use \code{Sys.Date()} i.e. today.
+#' @param min_age,max_age optional min and/or max dates that the Date of Birth could take as the century needs to be guessed.
+#' Must be either length 1 for a 'fixed' age or the same length as \code{chi_number} for an age per CHI number.
+#' min_age can be age based on common sense in the dataset, whilst max_age can be age when an event happens such as the age at discharge.
+#' @param chi_check logical, optionally skip checking the CHI for validity which will be
+#' faster but should only be used if you have previously checked the CHI(s), the default (TRUE) will to check the CHI numbers.
+#'
+#' @return an integer vector of ages in years truncated to the nearest year. It will be the same length as \code{chi_number}.
+#' @export
+#'
+#' @examples
+#' age_from_chi("0101336489")
+#'
+#' library(tibble)
+#' library(dplyr)
+#' data <- tibble(chi = c(
+#'   "0101336489",
+#'   "0101405073",
+#'   "0101625707"
+#' ), dis_date = as.Date(c(
+#'   "1950-01-01",
+#'   "2000-01-01",
+#'   "2020-01-01"
+#' )))
+#'
+#' data %>%
+#'   mutate(chi_age = age_from_chi(chi))
+#'
+#' data %>%
+#'   mutate(chi_age = age_from_chi(chi, min_age = 18, max_age = 65))
+#'
+#' data %>%
+#'   mutate(chi_age = age_from_chi(chi,
+#'     ref_date = dis_date
+#'   ))
+age_from_chi <- function(chi_number, ref_date = NULL, min_age = 0, max_age = NULL, chi_check = TRUE) {
+
+  # Do type checking on the params
+  stopifnot(typeof(chi_number) == "character")
+
+  if (!is.null(ref_date) & !inherits(ref_date, c("Date", "POSIXct"))) {
+    stop("ref_date must have Date or POSIXct class")
+  }
+
+  # min and max ages are in a reasonable range
+  stopifnot(min_age >= 0)
+
+  if (!is.null(max_age)) stopifnot(max_age >= min_age)
+
+  if (is.null(ref_date)) ref_date <- Sys.Date()
+
+  max_date.age <- ref_date - lubridate::years(min_age)
+
+  if (is.null(max_age)) {
+    min_date.age <- NULL
+  } else {
+    min_date.age <- ref_date - lubridate::years(max_age)
+  }
+
+  guess_dob <- dob_from_chi(
+    chi_number = chi_number,
+    min_date = min_date.age,
+    max_date = max_date.age,
+    chi_check = chi_check
+  )
+
+  guess_age <- age_calculate(guess_dob, ref_date)
```

Comment:
From my reading of `age_calculate` isn't the parameter `date_class` required? (see my other comment on `date_class`